In [2]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/creditcardfraud/creditcard.csv


In [5]:
df = pd.read_csv('/kaggle/input/creditcardfraud/creditcard.csv')
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [10]:
print(df.shape)
print(df.size)

(284807, 31)
8829017


<h2> Observation </h2>
<br> Rows = 284806 <br>
Columns/ Features = 31

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [13]:
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

<h2> Observation</h2>
Data contains no null values in the columns

### **Observation**
- Total columns/features 31
- Data contain total 284806 rows
- Time means the time difference from first transation
- Class column
    - 0 = transaction is legit,
    - 1= transation is frauds

In [14]:
#checking distribution of legit transation and fraudulent transaction
df['Class'].value_counts()

Class
0    284315
1       492
Name: count, dtype: int64

### **Observation**
- data distribution is highly imbalanced as legit transation count is 284315 where fraudulent transaction count is very less.


In [16]:
#separating the data for analysis
legit = df[df.Class == 0]
fraud = df[df.Class == 1]

In [17]:
print(legit.shape)
print(fraud.shape)

(284315, 31)
(492, 31)


In [19]:
legit['Amount'].describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [20]:
fraud['Amount'].describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [24]:
df.groupby('Class')['Amount'].mean().round(3)

Class
0     88.291
1    122.211
Name: Amount, dtype: float64

### Observation
- mean of legit transaction is 88
- where fraud transaction mean is 122
- mean of normal transaction is very different from mean of fraudelent transaction

## Dealing with unbalanced data
99.9% data is legit and very small amount of data is fraudulent  
**Under-Sampling Technique**

#Build a sample dataset containing simillar distribution of normal transaction(legit) and fraudelent transaction  
- Number of fraudelent transaction is 492
- Create a sample dataset from legit transaction using **Random Sampling** and join it with fraudelent dataset to make uniform distribution

In [25]:
legit_sample =legit.sample(n=492)
legit_sample.shape

(492, 31)

concatenating both data frame  
- legit_sample (492, 31)
- fraud(492, 31)

In [28]:
new_df = pd.concat([legit_sample,fraud], axis= 0)
print(new_df.shape)
print(new_df.head())

(984, 31)
            Time        V1        V2        V3        V4        V5        V6  \
74631    55676.0  1.263267 -0.740464  0.424465 -0.465971 -0.961485 -0.331113   
206947  136454.0  0.282089  0.828537 -0.585574  0.867551  0.836342 -0.448920   
187408  127521.0 -0.270179  0.588291 -1.146552 -0.711672  0.298838  1.723629   
60621    49411.0 -0.647479  1.083078  1.525104  2.488139  0.291960  0.526664   
122232   76485.0  1.460021 -1.136479  0.765496 -1.214282 -1.992043 -1.089946   

              V7        V8        V9  ...       V21       V22       V23  \
74631  -0.568037 -0.042307 -0.455096  ...  0.033721  0.123255 -0.043066   
206947  1.239483 -0.346021 -0.895493  ...  0.222509  0.761558 -0.161697   
187408  1.020325 -1.479699 -0.191328  ...  1.778063  0.375955 -0.150695   
60621   0.697690  0.246735 -1.180050  ...  0.012347  0.027197 -0.174773   
122232 -1.175811 -0.161541 -1.564488  ...  0.005544  0.371925 -0.016922   

             V24       V25       V26       V27       V28  

In [29]:
new_df['Class'].value_counts()

Class
0    492
1    492
Name: count, dtype: int64

In [32]:
new_df.groupby('Class')['Amount'].mean().round(3)

Class
0     84.894
1    122.211
Name: Amount, dtype: float64

**Observation**
- mean value is similar to the population mean value

<h1> Feature Engineering </h1>

In [35]:
#importing model and train_test_split

from sklearn.model_selection import train_test_split   #importing train_test_split for spliting data
from sklearn.linear_model import LogisticRegression    #importing LogisticRegression as the problem is classification problem
from sklearn.metrics import accuracy_score             #importing performance matrics accury_score to check model accuracy

In [36]:
X = new_df.drop(columns = 'Class', axis = 1)
Y= new_df['Class']

In [37]:
print(X)

            Time        V1        V2        V3        V4        V5        V6  \
74631    55676.0  1.263267 -0.740464  0.424465 -0.465971 -0.961485 -0.331113   
206947  136454.0  0.282089  0.828537 -0.585574  0.867551  0.836342 -0.448920   
187408  127521.0 -0.270179  0.588291 -1.146552 -0.711672  0.298838  1.723629   
60621    49411.0 -0.647479  1.083078  1.525104  2.488139  0.291960  0.526664   
122232   76485.0  1.460021 -1.136479  0.765496 -1.214282 -1.992043 -1.089946   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  .

In [38]:
print(Y)

74631     0
206947    0
187408    0
60621     0
122232    0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64


In [39]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, stratify=Y, random_state=2)  #spliting randomly

In [40]:
print('Shape of X_train', X_train.shape)
print('Shape of Y_train', Y_train.shape)
print('Shape of X_test', X_test.shape)
print('Shape of Y_test', Y_test.shape)

Shape of X_train (787, 30)
Shape of Y_train (787,)
Shape of X_test (197, 30)
Shape of Y_test (197,)


<h3> We use logistic regression as we have a classification problem </h3>

In [44]:
model = LogisticRegression()

In [45]:
model.fit(X_train,Y_train)

/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [46]:
#Model Evaluation
X_train_predict = model.predict(X_train)
Training_data_accurary = accuracy_score(X_train_predict, Y_train)
Training_data_accurary

#printing accuracy on training data
print("Training Data Accuracy is", Training_data_accurary)

Training Data Accuracy is 0.951715374841169


In [49]:
X_test_pred = model.predict(X_test)
Testing_data_accuracy = accuracy_score(X_test_pred, Y_test)
print("Testing Data Accuracy is ", Testing_data_accuracy)

Testing Data Accuracy is  0.9187817258883249


In [51]:
results = pd.DataFrame({
    "Actual": Y_test,
    "Predicted": X_test_pred
})

In [52]:
print(results.head())

        Actual  Predicted
247995       1          1
208135       0          0
187741       0          0
281734       0          0
15451        1          1


## **Conclusion**

In this ABC Bank Credit Card Fraud Detection, we combated class imbalance by balancing the dataset through sampling. Logistic regression achieved high accuracy (94% on training, 91.2% on test), demonstrating its effectiveness for fraud detection. This project serves as a strong basis for a robust credit card fraud detection system, with potential for further precision and recall optimization in real-world applications.
